magento/pwa-studio
# [feature]: Lazy Load/trim unused bytes in main routes

## Description

Our main `client` and `vendor` bundles are pretty large. There are savings (**130k split from this PR alone!**) to be had by splitting code that is unnecessary for the first render. That would include things like modals, drawers, drop downs, etc. The majority of changes in this PR are just adding `React.lazy` to components that aren't necessary on first render, and wrapping their implementations in `<Suspense>`.

Note: Much of the remaining unused bytes are from dependencies that we must load at start, such as Apollo, but for whatever reason do not tree-shake. Other bytes are from our implementation of ContextProviders which expose a lot of unnecessary actions in the main bundle. Further work is necessary to trim out that deprecated code. Definitely more work than is in the scope of this specific ask.

Testing these changes proved more difficult than I would have liked. Kent Dodds has [a way](https://awesomereact.com/playlists/dev-tips-with-kent/lfb5jvHq9c4) around testing Suspense/Lazy but we don't have react-testing-library at our disposal right now. The current approach (mocking `React.lazy`) is a stop-gap.

## Related Issue





Closes [PWA-1109](https://jira.corp.magento.com/browse/PWA-1109).

## Acceptance



### Verification Stakeholders

### Specification



### Verification Steps


1. Go to the home page/route.
2. "Clear site data" in dev tools.
3. Run the [coverage](https://developers.google.com/web/tools/chrome-devtools/coverage) tool.
4. Compare numbers to those on the same route/page on develop.pwa-venia.com. They should be **better**.
5. Repeat step 1-4 for all of the routes including: cms, category, product page, cart, checkout and my account.

## Screenshots / Screen Captures (if appropriate)

**Develop Category Page**
![](https://i.gyazo.com/cc535974abf44d5575511c3216860589.jpg)
**"Improved" Category Page**
[![Image from Gyazo](https://i.gyazo.com/1c198242e88b58c6658ea1f06de5fe52.jpg)](https://gyazo.com/1c198242e88b58c6658ea1f06de5fe52)

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.
